src/IECoreScene/CurvesAlgoUpdateEndpointMultiplicity.cpp

```diff
@@ -46,110 +46,74 @@ using namespace Imath;
+// the end by "adjust", either adding or removing values at both ends, depending on the sign of
+// "adjust"
 template<typename T>
-void expandVertex( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, const PrimitiveVariable &primVar, const IECore::CubicBasisf &targetBasis )
+void adjustEndPoints( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, PrimitiveVariable::Interpolation interpolation, int adjust )
 {
 	size_t numCurves = curves->numCurves();
-	out.reserve( in.size() + numCurves * 4 );
+	out.reserve( in.size() + numCurves * 2 * adjust );
 
-	size_t index = 0;
+	size_t curveIndex = 0;
 	for( size_t i = 0; i < numCurves; ++i )
 	{
-		size_t numSegments = CurvesPrimitive::numSegments( targetBasis, curves->periodic(), curves->variableSize( PrimitiveVariable::Vertex, i ) + 4 );
+		size_t numIn = curves->variableSize( interpolation, i );
 
-		// duplicate the first 2 values
-		out.push_back( in[index] );
-		out.push_back( in[index] );
-
-		for( size_t j = 1; j < numSegments; ++j, ++index )
+		if( adjust < 0 )
 		{
-			out.push_back( in[index] );
+			for( size_t j = -adjust; j < numIn + adjust; ++j )
+			{
+				out.push_back( in[curveIndex + j] );
+			}
 		}
+		else
+		{
+			// duplicate the start point
+			for( int j = 0; j < adjust; j++ )
+			{
+				out.push_back( in[curveIndex] );
+			}
 
-		// duplicate the last 2 values (note we've already incremented index for the next curve)
-		out.push_back( in[index-1] );
-		out.push_back( in[index-1] );
-	}
-}
-
-// remove the replicated first and last values reversing the 'expand' function above
-template<typename T>
-void compressVertex( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, const PrimitiveVariable &primVar, const IECore::CubicBasisf &targetBasis )
-{
-	size_t numCurves = curves->numCurves();
-	out.reserve( in.size() - numCurves * 4 );
-
-	size_t index = 0;
-	for( size_t i = 0; i < numCurves; ++i )
-	{
-		size_t numSegments = CurvesPrimitive::numSegments( targetBasis, curves->periodic(), curves->variableSize( PrimitiveVariable::Vertex, i ) - 4 );
-
-		// skip the first two values
-		index += 2;
+			for( size_t j = 0; j < numIn; ++j )
+			{
+				out.push_back( in[curveIndex + j] );
+			}
 
-		for( size_t j = 0; j < numSegments + 1; ++j, ++index )
-		{
-			out.push_back( in[index] );
+			// duplicate the end point 
+			for( int j = 0; j < adjust; j++ )
+			{
+				out.push_back( in[curveIndex + numIn - 1] );
+			}
 		}
 
-		// skip the last two values
-		index += 2;
+		curveIndex += numIn;
 	}
 }
 
-template<typename T>
-void expandVarying( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, const PrimitiveVariable &primVar, const IECore::CubicBasisf &targetBasis )
+// Set the endpoint multiplicities for vertex and varying primvars, based on the given basis
+void setMultiplicity( const IECore::CubicBasisf &cubicBasis, int &vertexMultiplicity, int &varyingMultiplicity )
```

Comment:
Or could it take both orig and new basis as args and return the final adjustment values instead? That'd reduce the call site from 6 lines down to 1.

Reply:
It's a separate function to avoid having to duplicate the if/else for mapping basis -> endPointMultiplicity for both input and output.

Having two outputs was a bit ugly though.  I've switched the whole thing up to use CurvesPrimitive::numSegments to derive varyingAdjustment from vertexAdjustment instead of manually specifying the varying values.  See if this feels better to you.

Reply:
Thanks, I do find it easier to follow now